jsdoc/services/transforms/extract-module-name.js

```diff
@@ -0,0 +1,45 @@
+// Match [module:]moduleName [REMINDER]
+var MODULE_NAME_REGEX = /^\s*((?:module:)?[a-zA-Z0-9\._/-]+)?\s*(.*)$/;
+
+/**
+ * Extract the name information from a module tag. If such information is not present
+ * it set the module name to its file projectRelativePath without the '.js' extension.
+ * It returns what is left after matching the module name (even if nothing should remain)
+ *
+ * @param  {Tag} tag The tag to process
+ */
+module.exports = function extractModuleNameTransform(log, createDocMessage) {
+  return function(doc, tag, value) {
+    if (tag.tagName == 'module') {
+
+      // JsDoc `@module` tags does not refer to any code node
+      // TODO: implement this a generic option for any tag to be declared inside tagDefs
+      //   E.g., `{ name: 'module', noCodeNode: true, transforms: [ ... ] }`
+      doc.codeNode = null;
+      doc.codeAncestors = null;
```

Comment:
I am worried that this will break other jsdoc tags...

For example, consider the following code:

```
/**
 * @module
 */

/**
 * Description of foo().
 */
  function foo() {}
```

Then I think that both comment blocks will be processed with respect to a single `doc` that has the `foo()` function as its `codeNode`.

Now the `@module` tag will cause this doc to lose connection to the `foo()` function.

The fact that we don't have a good way to distinguish JSDOC comments that should apply to the file only, is one reason why I have not properly implemented this before.